feat: migrate to uv, expand tests, CI/CD, and modernizations from textract-py3

KyleKing opened the pull request:
Hi! FWIW I have been maintaining a very minimal published fork of `textract` to address the blocking issue with `*` dependencies. This PR is to track the differences if anyone wants to use `textract-py3` while the long term maintenance of `textract` is resolved (#498). Once `textract` is released with a patch for #461, I will redirect users away from my fork because hopefully the upstream package will resume merging PRs, patching, and new version releases, but I unfortunately won't have the bandwidth to contribute beyond this fork.

Fixes: #128
Fixes: #184
Fixes: #279
Fixes: #324
Fixes: #375  
Fixes: #390
Fixes: #414
Fixes: #418
Fixes: #433
Fixes: #461
Fixes: #465
Fixes: #498
Fixes: #513
Fixes: #515
Fixes: #546
Fixes: #552

TheElementalOfDestruction committed:
Fix issue deanmalmgren#342

Clarification, _getStringStream *should* return `unicode` in Python 2, `str` in Python 3, IF the stream requested exists. If it does not exist, it returns `None`, which cannot be added to bytes. This commit adds a check for None, returning an empty bytes string if matched.

timgates42 committed:
docs: Fix a few typos

There are small typos in:
- docs/installation.rst
- textract/exceptions.py

Fixes:
- Should read `suppressed` rather than `supressed`.
- Should read `documentation` rather than `documenation`.
- Should read `accommodated` rather than `accomodated`.

Signed-off-by: Tim Gates <tim.gates@iress.com>

KyleKing committed:
build: initialize poetry (#1)

KyleKing committed:
docs: fix changelog version (#2)

KyleKing committed:
release: cleanup 2.0.1 release (#3)

LoicGrobol committed:
Enable encoding detection for the txt parser

As of now, the txt parser reads files in text mode as UTF-8 and fails with other encodings. This makes it return a bytes object, leaving the base `decode` to figure out the encoding and act accordingly.

KyleKing committed:
docs: remove reference to rtx

dhrim committed:
treat ShellError when call pdf2txt.py

branchv committed:
support python3.12

KyleKing committed:
Merge pull request #4 from deanmalmgren/master

docs: sync latest upstream changes

KyleKing committed:
Merge pull request #7 from timgates42/bugfix_typos

KyleKing committed:
Merge pull request #9 from branchvincent/py312

KyleKing committed:
Merge pull request #6 from LoicGrobol/patch-1

KyleKing committed:
Merge pull request #8 from TheElementalOfDestruction/fix-msg-support

KyleKing committed:
Merge pull request #10 from dhrim/master

KyleKing committed:
build: update dependencies

KyleKing committed:
Bump version: 2.0.1 → 2.1.0

KyleKing committed:
docs: add publish step to README

KyleKing committed:
Merge pull request #11 from KyleKing/release

StevenMapes commented:
@KyleKing and chance you could update the deps on your fork please and release a new version. It's pinned to using a older version of msg-extractor which had a pinned dep that broke when BS4 was updated, they've patched their part now, please see 
https://github.com/TeamMsgExtractor/msg-extractor/issues/450

Great work on the fork though keep it going please!

KyleKing commented:
Of course! I’m glad that you’re finding it to be useful and I’m happy to accept PRs. I’m out of town now, but I’ll release no later than the 24th!

KyleKing committed:
build: update lockfile

KyleKing commented:
My fork doesn't have any upper bound constraints on dependencies (https://github.com/KyleKing/textract-py3/blob/f3f509f8c36be05c85574ac841632693c2a19bd9/pyproject.toml#L22), so any project with `Python >=3.8,<4.0` can install `extract-msg==0.53.1`

For `textract-py3`, I don't plan on dropping support of Python 3.7 yet, but the relaxed specifications shouldn't introduce artificial version limits

anthonyhashemi commented:
Hey @KyleKing thanks for creating the fork and maintaing that for the moment. 

I tried using it here for 
but my colleague pointed out that the reason one of our tests started to break was because textract-py3 was installing . I fixed it by explcitly requiring `xlrd==1.2.0`.

When you get the chance you may want to restrict xlrd in the toml to `xlrd = ">=1.2.0,<2.0.0"` (I think)

KyleKing commented:
Thanks @anthonyhashemi! You're right and I'll publish a release this afternoon

Long-term, maybe adapting changes to use `openpyxl` for `xlsx` from #433 would be worthwhile in the case that `xlrd` no longer works reliably for `xlsx`

KyleKing committed:
fix: prevent errors parsing xlsx

Prevents upgrading xlrd

KyleKing commented:
I've released v2.1.1 with the patch. Thanks again!

https://pypi.org/project/textract-py3

KyleKing committed:
build: finish updating release documentation

StevenMapes commented:
Moving over to use openpyxl for xlsx would be good as then the constraint on xlrd could be removed which would be useful for projects that also use pandas for XLS as that, pandas, now requires v2.0.1+ of xlrd when using the xlrd engine

KyleKing commented:
That would be great and I would be open to PRs! I don’t have a use case for Excel, so I wouldn’t make the changes otherwise

KyleKing committed:
build: migrate to uv

KyleKing committed:
ci: initialize CI

KyleKing committed:
feat!: cleanup documentation and migration to preferred tooling

KyleKing committed:
refactor: init ruff, pyright, and run initial format

KyleKing committed:
ci: unsafe fixes with ruff

KyleKing committed:
refactor: additional round of fixes

KyleKing committed:
ci: add noqa for now

KyleKing committed:
refactor: additional cleanup and modernization

KyleKing committed:
fix: resolve escaping to address blocking conda error

KyleKing committed:
fix: parse first image in Gifs for tesseract OCR

KyleKing committed:
docs: add Brewfile

KyleKing committed:
test: resolve failing tests

KyleKing committed:
docs: merge READMEs

KyleKing committed:
refactor: migrate to uv-build and commitizen

KyleKing committed:
feat: add trusted publishing

KyleKing committed:
test: add dependencies in CI to improve coverage

KyleKing committed:
ci: fix versioning and dependency issues

KyleKing committed:
fix: resolve Windows failures

KyleKing committed:
fix: restore cross-platform

KyleKing committed:
ci: potential fixes

KyleKing committed:
fix: address beautifulsoup warnings and CI flakiness

KyleKing committed:
ci: comprehensive error reporting on each run

KyleKing committed:
ci: attempt to fix most recent Windows failures

KyleKing committed:
test: fix recent Windows failures

KyleKing committed:
test: explore using Act locally

KyleKing committed:
test: add improve debugging output on CI failures

KyleKing committed:
test: skip failing Windows tests

Root Cause

1. UTF-8 character handling: Windows versions of pdftotext and pdfminer
convert UTF-8 characters to ASCII equivalents:
    - Expected: b'that\xe2\x80\x99s' (UTF-8 right single quote ')
    - Actual: b"that's" (ASCII apostrophe ')
2. Non-breaking spaces: pdfminer on Windows outputs replacement
characters (\xef\xbf\xbd) for UTF-8 non-breaking spaces (\xc2\xa0)
3. Line count differences: Windows handles tabs, CR, and whitespace
differently, resulting in 29 vs 63 lines

KyleKing committed:
build: update minimum dependencies

KyleKing committed:
docs: cleanup top-level markdown files

KyleKing committed:
bump: version 2.1.1 → 3.0.0rc0

KyleKing committed:
docs: update license

KyleKing committed:
fix: prevent Path.startswith error

KyleKing committed:
refactor: run ruff

KyleKing committed:
bump: version 3.0.0rc0 → 3.0.0

KyleKing committed:
Merge branch 'master'

KyleKing commented:
Update: working with Dean on incrementally merging improvements from textract-py3 into textract over the next few weeks!

KyleKing committed:
refactor!: rename to textract to prepare to merge

KyleKing committed:
doc: notes on fixing syncing

KyleKing committed:
refactor: undo unrelated changes

KyleKing committed:
refactor: undo changes to tests

KyleKing committed:
refactor: undo additional changes

KyleKing committed:
refactor: reduce doc differences, remove mise/brewfile

KyleKing committed:
refactor: restore 1.6.4 version

KyleKing committed:
fix: restore patches from -py3

KyleKing committed:
Merge remote-tracking branch 'upstream/master'

KyleKing commented on .github/workflows/ci_pipeline.yml:
```diff
@@ -0,0 +1,84 @@
+      - name: Run tests
+        id: tests
+        continue-on-error: true
+        run: uv run pytest
```
CI is passing with latest push: https://github.com/KyleKing/textract-py3/actions/workflows/ci_pipeline.yml

KyleKing commented on .github/workflows/ci_pipeline.yml:
```diff
@@ -0,0 +1,84 @@
+      - name: Build Package
+        run: uv build
+      - name: Publish to PyPI
+        uses: pypa/gh-action-pypi-publish@release/v1
```
@deanmalmgren, requires someone with admin access to PyPi to go here and set the below values: https://pypi.org/manage/project/textract/settings/publishing

- Owner: `deanmalmgren`
- Repository: `textract`
- Workflow: `ci_pipeline.yml`
- Environment: `pypi`

deanmalmgren replied:
✅ ok. i have set this up in pypi now

❓ @KyleKing am I understanding correctly that this will create a PyPi release anytime we create a PR with passing tests on main that has a tag? Pretty slick, but I want to make sure I am understanding this mechanism correctly

KyleKing replied:
That's correct! Pushing the tag will trigger the publish action, which will release to PyPi and publish a GitHub Release if CI passes. I'm likely to merge bump_version.yml into this release step to also push the updated Changelog, but I'll think about your feedback below first!

deanmalmgren replied:
Sounds awesome. Thanks for setting this up 👏

KyleKing commented on docs/conf.py:
```diff
@@ -1,271 +1,44 @@
-# -*- coding: utf-8 -*-
```
This file was outdated, so I removed the boilerplate and just kept the settings that were still applicable

KyleKing commented on docs/contributing.rst:
```diff
@@ -3,124 +3,104 @@
+   The primary contributor documentation is in `CONTRIBUTING.md <https://github.com/deanmalmgren/textract/blob/main/CONTRIBUTING.md>`_
+   at the repository root. This page provides a summary for the documentation site.
 
+This is a maintained fork of `deanmalmgren/textract <https://github.com/deanmalmgren/textract>`_
```
TODO: I need to finish cleaning up this file

KyleKing commented on docs/installation.rst:
```diff
@@ -78,52 +83,38 @@ Ubuntu / Debian while using ``pkg`` as package manager.
-   of the project with the lower-cased name of your operating system
-   (e.g. ``requirements/windows``) so we can try to keep these things
-   up to date in the future.
+Installing system packages is a bit of a drag and it's hard to
```
TODO: And clean up this one ^

KyleKing commented on provision/debian.sh:
```diff
@@ -1,20 +0,0 @@
-# Install all of the dependencies required in the examples.
-# http://docs.travis-ci.com/user/installing-dependencies/#Installing-Ubuntu-packages
-apt-get update -qq
-sed 's/\(.*\)\#.*/\1/' < $base/requirements/debian | xargs apt-get install -y --fix-missing
```
These files are no longer needed with GitHub Actions

KyleKing commented on requirements/python-doc:
```diff
@@ -1,5 +0,0 @@
-
-sphinx==2.1.2
-sphinx_rtd_theme==0.4.3
-sphinx-argparse==0.2.5
```
Now defined with uv

KyleKing commented on scripts/prepend-changelog.sh:
```diff
@@ -0,0 +1,91 @@
+# Usage:
+#   ./scripts/prepend-changelog.sh        # After cz bump, adds the new version section
+#   ./scripts/prepend-changelog.sh --preview  # Preview unreleased changes before bumping
+
```
This script is optional and run in CI when publishing to ensure that the Changelog is up to date with all changes, relevant contributors, etc. as long as the PR title on squash is updated

I don't mind removing this and keeping the changelog up to date manually instead and don't have strong feelings

KyleKing commented on tests/jpg/standardized_text.txt:
```diff
@@ -0,0 +1,5 @@
+
+jumps over
+
+the lazy dog
```
Adding these files to git seemed to be required to reliably get tests working locally and in CI

deanmalmgren replied:
makes good sense to me

KyleKing commented on tests/base.py:
```diff
@@ -1,217 +1,308 @@
-import requests
+import textract
+
+
```
TODO: I made a number of changes to the test suite to get it running again. I'll plan to take another look to make sure I haven't made any unintentional changes

KyleKing commented on tests/test_doc.py:
```diff
@@ -1,7 +1,23 @@
+    not _HAS_ANTIWORD,
+    reason=(
+        "antiword is not installed "
+        "(no longer maintained, install via: brew install antiword)"
```
Long-term: likely need to replace with abiword, but requires additional investigation (see #468)

KyleKing commented on tests/test_exceptions.py:
```diff
@@ -1,57 +1,69 @@
     """This class contains a bunch of tests to make sure that textract
     fails in expected ways.
-    """
+    """  # noqa: D205
```
TODO: Remove unused ruff ignores

KyleKing commented on textract/bin/textract.py:
```diff
@@ -1,24 +1,18 @@
-#!/usr/bin/env python
```
Not necessary now that this is an entrypoint managed by uv

KyleKing commented on textract/parsers/__init__.py:
```diff
@@ -2,12 +2,12 @@
 import re
 
-from .. import exceptions
+from textract import exceptions
```
Ruff switched these to absolute imports. I'm not sure if you have a strong preference either way, but ruff still sorts them in the right order

deanmalmgren replied:
[dust] I would prefer them to be relative imports but this is not a hill I'm going to die on

KyleKing replied:
Yeah, ruff only has TID252 ("Prefer absolute imports over relative imports from parent modules" - https://docs.astral.sh/ruff/rules/#flake8-simplify-sim) and doesn't support pure relative imports without adding a second linter package. My preference would be to stick with what is auto-fixable, if you don't feel strongly?

KyleKing commented on textract/parsers/audio.py:
```diff
@@ -59,6 +64,6 @@ def convert_to_wav(self, filename):
-        temp_filename = '{0}.wav'.format(self.temp_filename())
-        self.run(['sox', '-G', '-c', '1', filename, temp_filename])
+        temp_filename = f"{self.temp_filename()}.wav"
+        self.run(["sox", "-G", "-c", "1", filename, temp_filename])
```
TODO: Remove double quote changes to make the actual changes more obvious

deanmalmgren replied:
all good. i prefer double quotes too

KyleKing commented on textract/parsers/gif_parser.py:
```diff
@@ -1 +1,54 @@
+        return stdout
+
+
+__all__ = ["Parser"]
```
This is net-new because it seemed to be expected by the tests, so I implemented it

KyleKing commented on CONTRIBUTING.md:
```diff
@@ -1,116 +1,290 @@
-   --- that should be self-evident from the code itself.
+# Contributing to textract
+
+Maintained fork of [deanmalmgren/textract](https://github.com/deanmalmgren/textract) focused on minimizing dependencies and supporting modern Python.
```
TODO: Also need to likely restore with only light changes to document usage of uv

KyleKing commented on LICENSE:
```diff
@@ -1,6 +1,7 @@
 The MIT License (MIT)
 
 Copyright (c) 2014 Dean Malmgren
+Copyright (c) 2023 Kyle King
```
I think this is right? Feel free to recommend changes

KyleKing commented on setup.py:
```diff
@@ -1,64 +0,0 @@
-    author_email='dean.malmgren@datascopeanalytics.com',
-    license='MIT',
-    scripts=scripts,
-    packages=[
```
Replaced with pyproject.toml

deanmalmgren replied:
amazing, thank you

KyleKing commented on Vagrantfile:
```diff
@@ -1,65 +0,0 @@
-
-    # these provisioning steps are only done locally as a convenience
-    # for setting up a useful development environment
-    provision_script(server_config, "provision/development.sh")
```
These have been migrated to GitHub Actions

deanmalmgren replied:
oh god. Vagrantfile. That was an era.

KyleKing reviewed (commented):
@deanmalmgren this is ready for a cursory review. Let me know:

1. Should this be split into smaller PRs?
1. Do you have any broad concerns with any of these changes? I took out what I thought might delay the PR (mise, Brewfile, Act), which could be added later, so I think what is left is generally good

FYI: I made progress on minimizing the changes between our repositories and left myself a few `TODO` comments so you can know what I'm still working on

KyleKing committed:
refactor: remove act, restore pyup (temporarily), and remove ruff config

KyleKing committed:
ci: fix bad merge

KyleKing committed:
docs: restore upstream documentation

KyleKing committed:
refactor: remove noqa's

KyleKing committed:
docs: minimize documentation changes

KyleKing committed:
test: manually review and patch test changes

KyleKing committed:
docs: update guidance for Windows and remove Docker

KyleKing committed:
docs: standardize on CI-based changelogs

KyleKing committed:
ci: move release to mdformat-style with manual trigger

KyleKing commented:
@deanmalmgren, I think the changes are now representative of what I think would be a good improvement to prepare textract for a 1.7.0 maintenance release! All of my 'todo' notes have been addressed and now it just depends on if this should be multiple PRs and/or if you have any concerns about the changes generally (https://github.com/deanmalmgren/textract/pull/543#pullrequestreview-3748830370) and then getting PyPi configured (https://github.com/deanmalmgren/textract/pull/543#discussion_r2762078955)

KyleKing committed:
docs: de-duplicate contributing documentation

deanmalmgren commented on .github/workflows/ci_pipeline.yml:
```diff
@@ -0,0 +1,60 @@
+
+      - name: Run tests
+        id: tests
+        continue-on-error: true
```
[dust] seems like we do not need to continue-on-error here if we exit on error in the next step?

KyleKing replied:
Removed! This was from when I was troubleshooting Windows

deanmalmgren commented on docs/changelog.rst:
```diff
@@ -5,11 +5,6 @@ This project uses `semantic versioning <http://semver.org/>`_ to
 track version numbers, where backwards incompatible changes
 (highlighted in **bold**) bump the major version of the package.
 
-
-latest changes in development for next release
-----------------------------------------------
-
-.. THANKS FOR CONTRIBUTING; ADD YOUR UNRELEASED CHANGES HERE!
```
[rock] This is the place where I'd typically aggregate changes from other users that were made during PRs before I created a new release for documentation purposes. I don't think we want to delete this unless you have a different proposal for how to accomplish this

KyleKing replied:
My thinking is that we would use `commitizen` to auto-generate this section based on commit messages. The proposed process would be:

1. Users open a PR (neither the contributor nor maintainer would need to push updates this section before merging, which saves time on checkout/edit/push and is the main benefit of this change)
2. Once approved, the maintainer squashes and updates the commit message and description as needed (using conventional commits, `fix: <>`, etc.)
3. When time to release, a maintainer locally runs commitizen (`uv run cz bump`) to update the version (based on conventional commits) and changelog, then they push the documentation changes and tag (`git push --tags`)

deanmalmgren replied:
@KyleKing I'm not as familiar with commitizen, but could we auto-generate this section based on the MERGE COMMIT messages rather than squashing? I don't particularly like the squash merge approach and have been burned by wanting previous commits that were squashed and now impossible to find.

KyleKing replied:
Oh okay, on merge commits, any commit in the history would be included in the changelog if it follows the conventional commit pattern, which we may not want

Commitizen and other automated tools kind of dictate how to define commits in main, so I think it might be better to keep it simple and keep manual versioning? I'll plan to remove commitizen before merging, but keep the GitHub Releases (which are automatically generated from PRs)

---

On Squash vs. Merge, do you have a preference for which style to use every time or is it case-by-case? For example, I was planning on generally favoring squash, but would you prefer that we favor merging? I would like to squash this PR in particular because there are a lot of noisy commits testing out different CI bugs/etc.

KyleKing replied:
Updated in https://github.com/deanmalmgren/textract/pull/543/commits/8334ebfc6d3d6b12934a2834421d67acd7bdcb25

deanmalmgren commented on tests/Makefile:
```diff
@@ -38,7 +38,7 @@ ps/raw_text.txt: ps/raw_text.ps
 
 %.txt: %.gif
-	f=$@; g=$${f%%.*}; tesseract $< $${g} > /dev/null
+	uv run python -c "import sys; import textract; sys.stdout.buffer.write(textract.process('$<'))" > $@
```
[rock] this is unexpected. If I remember correctly, the Makefile is used to create the test data that is used to verify the output. It's a little strange to allow textract to be used to create the answers to it's own tests

KyleKing replied:
Fixed!

deanmalmgren commented on tests/run.py:
❓ RIP Travis? I'm fine with that. I just want to make sure I understand the workflow

KyleKing replied:
Yes! GitHub Actions now runs everything that would have been run in Travis and handles cd and other setup without this wrapper script

deanmalmgren commented on tests/test_gif.py:
```diff
@@ -1,7 +1,20 @@
+
+@pytest.mark.skipif(
+    not _HAS_TESSERACT,
+    reason="tesseract-ocr is not installed (install via: brew install tesseract)",
```
[dust] would be better to reference install instructions for different operating systems.

KyleKing replied:
Updated!

deanmalmgren commented on tests/test_html.py:
```diff
@@ -1,18 +1,83 @@
+        reason=(
+            "HTML output varies on Windows "
+            "due to BeautifulSoup/lxml platform differences"
+        ),
```
oy ve. that is super annoying. 

💡 I certainly won't block this PR from being merged, but if this is a pattern we run into frequently, we may want to consider switching to a pattern where we separately maintain test data for different operating systems if this comes up frequently. Again, I am not expecting us to do this right now, but it's a pattern we could start to use if system-specific differences are an issue.

KyleKing replied:
I'm going to take a look and see if there is an easy fix

KyleKing replied:
Fixed!

deanmalmgren reviewed (commented):
@KyleKing this is amazing. Thanks for doing the work to aggregate everything here. I like to use [Netlify's feedback ladder](https://www.netlify.com/blog/2020/03/05/feedback-ladders-how-we-encode-code-reviews-at-netlify/) for PR feedback. Dust feedback is "take it or leave it" and rock feedback is more significant, potentially blocking changes. I have a couple of clarifying questions (❓), mostly around what the workflow looks like now. 

I can't tell you how grateful I am to be moving out of the Vagrant era of textract development into a uv+github actions era. Much appreciated!

KyleKing committed:
ci: remove continue on error

KyleKing committed:
test: restore non-python generated test files

KyleKing committed:
test: generalize guidance on installation when missing for pytest

KyleKing committed:
build: add missing lxml

KyleKing committed:
docs: restore read the docs generation

KyleKing committed:
ci: run ruff

KyleKing commented on .readthedocs.yaml:
```diff
@@ -0,0 +1,19 @@
+# Read the Docs configuration file for Sphinx projects
+# See https://docs.readthedocs.io/en/stable/config-file/v2.html for details
+
+version: 2
```
No changes should be necessary, but after merging, we should confirm that the RTD action works

KyleKing reviewed (commented):
The last two months have been busy (my brother visited, we had some close friends leaving post, and I worked a number of long days to wrap up projects at work). Our son is doing well and I'm on paternity now, so I conversely actually have some time to get back to this!

deanmalmgren commented:
@KyleKing First and foremost, congrats on parenthood! Hope you and your growing family are doing well :)

The changes here look great. What remains to be done here before we can merge this in?

KyleKing committed:
ci: merge release into the CI pipeline

KyleKing committed:
test: output bytes diff as well

KyleKing committed:
fix: likely resolve differences across OS

KyleKing committed:
test: remove temporary debugging in CI

KyleKing committed:
ci: cleanup temporary lxml OS workarounds

KyleKing commented:
Thanks! The last thing I was looking into were differences across Windows and gaps in that testing, which should now be mostly complete with one last investigation to see if there is an easy fix for PDF

KyleKing committed:
ci: expand testing on Windows

KyleKing committed:
ci: fix choco installations

KyleKing committed:
fix: use Windows shell

KyleKing committed:
fix: resolve issues with Windows

KyleKing committed:
test: further investigate Windows issues

KyleKing committed:
fix: resolve last Win CI test failure?

KyleKing committed:
refactor: (nit) consistent usage of Pathlib

KyleKing committed:
test: fix `make all|force` and re-generate test files

KyleKing committed:
test: rewrite test_pdf

KyleKing committed:
ci: use templates for committizen

KyleKing committed:
ci: simplify release process

KyleKing committed:
test: migrate to functional-style

KyleKing committed:
test: finish unittest style conversion

KyleKing committed:
ci: add Dependabot for Actions

KyleKing committed:
test: resolve lints and CI failure

KyleKing committed:
refactor: add maintainers

KyleKing committed:
refactor: revert out of scope stylistic changes

KyleKing committed:
test: backport improvements

KyleKing committed:
ci: remove outdated pyup config

KyleKing committed:
ci: simplify choco installs

KyleKing committed:
ci: resolve ruff, pyright, and pytest

KyleKing committed:
fix: remove six workaround

KyleKing committed:
build: partially address CVEs

KyleKing committed:
ci: fix Mac 3.9 failure

KyleKing commented on .pyup.yml:
Do you still want to use pyup with uv? Or use Dependabot or just update manually? My preference is for Dependabot or manual because the changes can be limited to security patches

deanmalmgren replied:
I also prefer dependabot. Thanks!

KyleKing commented on textract/parsers/html_parser.py:
```diff
@@ -13,42 +14,72 @@ class Parser(BaseParser):
-            text = re.sub(r'(<[^>]+>)', '', text)
-            text = re.sub(r'\s', ' ', text)
+            text = re.sub(r"(<[^>]+>)", "", text)
+            text = text.replace("\r\n", "\n").replace("\r", "\n")
```
FYI: added to fix Windows compatibility

KyleKing commented on textract/parsers/__init__.py:
```diff
@@ -86,21 +89,20 @@ def _get_available_extensions():
+    # Escape the path for regex to handle Windows backslashes and special chars
+    ext_re = re.compile(
+        re.escape(glob_filename).replace(re.escape("*"), r"(?P<ext>\w+)"),
+    )
```
Additional changes required to expand Windows consistency

KyleKing committed:
refactor: last cleanup!

KyleKing commented on tests/test_exceptions.py:
```diff
@@ -12,46 +18,50 @@ class ExceptionTestCase(base.GenericUtilities, unittest.TestCase):
+                check=False,
+            )
+        finally:
+            Path(filename).unlink(missing_ok=True)
```
As part of finishing the Python 3 migration, I removed os in favor of pathlib

deanmalmgren replied:
Amazing; thanks!

KyleKing committed:
refactor: finish removing usage of six

KyleKing committed:
fix: support pocketsphinx on MacOS when no available wheel

KyleKing reviewed (commented):
@deanmalmgren, ready for review again! I think I've addressed the feedback, improved CI testing of Windows, and I'm finally happy with this set of changes. Once merged, I'll make a major release candidate (rc) release to PyPi to test that everything works, then I can make the official release

KyleKing commented:
@deanmalmgren, this PR is ready for review. Let me know if you have any concerns/questions and, if easier, I can also do a sync call

deanmalmgren commented on Makefile:
```diff
@@ -0,0 +1,41 @@
+
+sync:
+	uv sync --all-extras
+	@if [ "$$(uname)" = "Darwin" ]; then \
+		find .venv -name "_pocketsphinx.cpython-*-darwin.so" -exec codesign -s - -f {} \; 2>/dev/null || true; \
+	fi
```
yuck. pity this is needed on osx.

deanmalmgren reviewed (approved):
BRAVO!!! Nicely done @KyleKing . 

I'm ready to merge if you are 🙌 💥 🙌

KyleKing committed:
feat: restore improved manual changelog process

KyleKing committed:
docs: update release notes for 2.0.0

KyleKing committed:
docs: don't forget the docs changes!

KyleKing commented:
I've [released v2.0.0](https://pypi.org/project/textract/2.0.0) with the new Trusted Publishers integration! A couple of minor issues:

1. I intended the release to be a release candidate (v2.0.0rc1) - I overlooked that the version is from the pyproject.toml/uv rather than the tag where I'm used to those being kept in-sync with commitizen
2. RTD isn't building - I have pre-releases turned off in the configuration

I'll follow up later today to close open issues, monitor for new issues, and generally continue verifying that v2.0.0 is working as intended

deanmalmgren commented:
amazing work @KyleKing ! Thanks for your collaboration here.
